src/index/coinstatsindex.cpp

```diff
@@ -228,10 +228,7 @@ bool CoinStatsIndex::WriteBlock(const CBlock& block, const CBlockIndex* pindex)
     m_muhash.Finalize(out);
     value.second.muhash = out;
 
-    CDBBatch batch(*m_db);
-    batch.Write(DBHeightKey(pindex->nHeight), value);
-    batch.Write(DB_MUHASH, m_muhash);
-    return m_db->WriteBatch(batch);
```

Comment:
I think this is spot is problematic, and the reason for the behavior reported in #24076: 

If we persist `DB_MUHASH` here in each `BlockConnected()` callback, but not `DB_BEST_BLOCK` (which is persisted only in `ChainStateFlushed()` callbacks and will therefore at this point in time have a lower height), the db is temporarily in an inconsistent state.

So in case of an unclean shutdown (with no time to flush the chainstate to disk to get in sync again), on next startup the index will initialize with the older `DB_BEST_BLOCK` but a `DB_MUHASH` which contains data from newer blocks. The index then reprocesses the blocks after `DB_BEST_BLOCK`'s height a second time. 
But MuHash is a rolling hash, so it will incorporate data from these blocks twice, and the hash will be incorrect.

Reply:
In commit "index: Commit DB_MUHASH and DB_BEST_BLOCK to disk together" (010156c07f28e2c8dc30d4408942d1e1d88ab54d)

> The `CommitInternal` override feels like a hacky refactor.

I think I agree with luke if the point is that the override-and-call-base class code in `CommitInternal` is a little hacky, but fjahr's right this just the way the base index class was designed, so the override seems like appropriate way to handle this now. #24230 does clean up all the overrides and will simplify this a bit.

Reply:
Just to make this clear: This is not a refactor, it changes behavior (the points in time when `DB_MUHASH` is written to disk).

My understanding is that the indexes work in general such that that all persisted data describing the current _progress_ of indexing (such as the best block) must be in sync with the persisted chainstate information (which is sometimes not flushed for several blocks received) - that is why `ChainStateFlushed()` callbacks are used. In case of an unclean shutdown, the chainstate will be constructed from this point on, so the indexes must as well.

In contrast to this is the actual indexed data for each block - I think we could in principle cache it and also write it to disk with each `ChainStateFlushed()` callback and not use `BlockConnected()` at all. 

But it is also possible to write it to disk in advance with each `BlockConnected()` callback, as it is done:
This is dependent on the assumption that we will not read (and possibly overwrite) the indexed data in case it is ahead of the flushed chainstate data (and the best block), and also that _before_ flushing the chainstate, we have called `BlockConnected()` for each block.

With this in mind, I don't see a better way of doing this other than overriding `CommitInternal`.